### PR TITLE
QIP - MessageTransaction Encoded Message Standard

### PR DESCRIPTION
**Pull request by scottdonaldau**
A standard message encoding format to indicate encoded data in MessageTransaction transactions on the QRL Network. Open for discussion.

### CODE REVIEW

**Comment by 0xFF0** [comment]
I do support the idea of a standard message encoding format. 
 
However, is there a reason why the encoded message needs to start by `AFAF` ?

With the current specification, the number of possible type of messages is 241 (i.e: 256 - 16 (can't use the hex string A) + 1 (document notarisation)).
 
Maybe, the encoded message could be implemented this way?
 
`AF XX [MSG_MAX_77_BYTES] YY`
 
- AF: indicate that it's an encoded message.
- XX: Indicate a unique encoded message transaction type (0xAF would be for document notarisation).
- YY: 1 byte checksum. 
 
With this format, there could be 256 possible types of messages (instead of 241). Also, the checksum byte could provide a more robust message. 
 
The checksum at the end of the message could be used to:
- Indicate that it’s an encoded message.
- Applications build on top of the MessageTransaction (like coinvote) could use this checksum for a more robust analysis.
 
Also, by adding a checksum, the possibility of having a “bad” encoded message would be lower (it’s more probable to have a message with AFAF compared to AF [...] YY).
 
 
Btw, I think [line 38] should be "The remaining seventy-seven"
 
[line 38]: https://github.com/theQRL/qips/blame/ce08b23212310a6eac5901d7d51f4ee762a951c1/2.Proposals/1.%20Open/2%20-%20MessageTransaction%20Encoded%20Message%20Standard/README.md#L38

**Comment by surg0r** [comment]
I think a checksum on the end of the message field is a nice idea, @0xFF0. I would guess @scottdonaldau wants to preserve the existing format for previous notarisation transactions and therefore wishes to keep `AFAF` as the message encoder.

It might be useful to scan the blockchain to look for messages which adhere to the notary format and if just a handful support both cases in the explorer. This would allow us to have a more 'perfect' message encoding standard without worrying about deprecating previous iterations.

**Comment by ikichagit** [comment]
KEEP V7

For a while, pls keep at v7, v8 is bad idea and why always following monero?
Or, make custom v7 algorithm like Masari, Stellite or Haven. This better than following monero

**Comment by surg0r** [comment]
I have adapted my coinvote scripts to search for every `message_tx` using `AFAF` as the initial encoding bytes such as in the case of notary transactions. Scanning the entire chain reveals `25` tx in total to current blockheight `196107`.

So perhaps @scottdonaldau we could alter the two initial encoding bytes to something else and incorporate a checksum, but also support the older format for display purposes (though move to newer format in the webwallet)?

**Comment by cyyber** [comment]
I like the idea by @0xFF0 . Although I believe that we can ignore AF, saving 1 byte, and may have the format as specified by @0xFF0 .

- XX: Indicate a unique encoded message transaction type (0xAF would be for document notarisation).
- YY: 1 byte checksum.

Thus before processing any txn, if XX matches with any encoded message type and YY checksum matches, only then process the encoded message. 

Although for document notarization we may have an exceptional case. But for the upcoming encoded messages, we can simply use above mentioned format.

**Comment by surg0r** [comment]
As discussed @cyyber I would think a single encoding byte would be wrong since there is only a `2^8` chance of random message data being incorrectly encoded.

**Comment by 0xFF0** [comment]
Thanks @surg0r and @cyyber .

I think the format with a checksum is more than `2^8` chance of being incorrectly encoded. 

Here's an example of what I was thinking of:

```
import hashlib

def isEncodedMessage(p_msg, p_blockHeight):
    if 'AF' in p_msg[0:2]:
        checksum = hashlib.sha256(p_msg[:-2]).hexdigest()[0:2]
        if checksum in p_msg[-2:]:
            return True
        elif p_blockHeight < 196107: #Assuming QIP done at block 196107
            if "AFAFA" in p_msg[0:5]:
                return True
				
    return False

#Old format
oldDocNotarisation = "AFAFA3" + "D41D8CD98F00B204E9800998ECF8427E" #MD5 empty doc

#New format
msg = "AF00" + "QRL".encode("hex")
valid_msg = msg + hashlib.sha256(msg).hexdigest()[0:2]

print isEncodedMessage(valid_msg, 200000) #True
print isEncodedMessage(msg, 200000) #False
print isEncodedMessage(oldDocNotarisation, 196000) #True
print isEncodedMessage(oldDocNotarisation, 200000) #False 
```

The chance of having a random message with `AFAF` is currently `2^16`

By adding a checksum and saving 1 byte (i.e: `AF XX [MSG_MAX_77_BYTES] YY`), the chance is far more > than `2^16` because of the hash function involved in the checksum format.

**Comment by scottdonaldau** [comment]
@0xFF0 Just to be clear here, this isn't a hard fork - this is purely a layer 2 standard format that will live inside a MessageTransaction. There is no block number considerations with respect to this QIP - simply defining the encoding structure so that apps like a block explorer or wallet can integrate seamlessly.

**Comment by cyyber** [comment]
Yes 2^8 will result into a lot of collision, but its important to decide if 2^16 is enough or do we need more bytes to avoid collision. 

Although checksum solves this problem, but once atleast first 2 bytes matches with any kind of message encoding format, checksum needs to be calculated which is only possible processing the full message. Since its a public blockchain, probably 2^16 might not be enough. May be 4 bytes to denote message encoding + 1 byte checksum, may fill the shortage for future purposes.

**Comment by 0xFF0** [comment]
Thanks @scottdonaldau 

I was not thinking about a hard fork. I thought that wallet/explorer did have a way to get the block height associated with the message (I have not looked at the code so sorry if I'm wrong about this). The idea of looking at block height was to keep a backward compatibility with the 25 transactions in the blockchain.

@cyyber If a streaming solution is needed (i.e: determine if it's an encoded message without processing the full message), I agree 2^8 is not enough. But don't you think that a checksum based on sha256 (or something similar) would not be enough?

**Comment by surg0r** [comment]
@scottdonaldau: using a `1 byte` encoding byte and a `1 byte checksum` (no matter how it is calculated - the hash function used doesn't alter the odds!) actually results in exactly `1 in 2^16` chance of a collision, with random data being incorrectly encoded as a valid message and passing checksum in the new format.

With a 2 prepended bytes encoding and a single checksum byte the risk of a collision and false positive encoding falls to `1 in 2^24 (16.77M)`, but the risk of the checksum being wrong (again, regardless of function used) is only `1 in 2^8 (256)`. 

Therefore, it makes a bit more sense to me to use a single prepended byte to denote our message format with a two byte checksum. Then, the risk of a random collision being encoded as valid in our message format is still tiny at `1 in 2^24 (16.77M)`. But the accuracy of the checksum for the payload `77 bytes` rises 256 fold with the chance of an error falling considerably to `1 in 2^16 (65536)`.

tldr: single prepended encoding byte and two byte checksum is better imho

**Comment by cyyber** [comment]
- Plan A (3 or 4 prepended bytes encoding and No hard-coded checksum)

Instead of hard coding checksum, we may leave that on user. The real advantage of not having hard coded checksum, is in case users doesn't care much about the correctness of his data, it can simply take leverage of all the remaining bytes to store the data. 
Moreover, if a user needs a checksum of 2 bytes or 4 bytes depending upon his own needs, he can simply add that checksum on message field of message transaction. 
Thus having no hard coded byte for checksum, gives user flexibility to have a variable length checksum which depending upon his needs.

- Plan B (3 or 4 prepended bytes encoding and 1 byte hard-coded checksum)

In case we want to hard code checksum, we can hard code 1 byte checksum. This will force user to have checksum of atleast 1 byte for all the messages, whether he needs it or not. Moreover, if user needs checksum of more than 1 byte, lets say 3 bytes checksum, then user may add 1 byte value of checksum in hard-coded checksum field, and rest 2 bytes in its own message. In such a way, user may achieve a variable length checksum where minimum length of checksum is 1 byte.


I prefer Plan A, as it is more cleaner and makes sense for all kinds of use cases.

**Comment by surg0r** [comment]
- Plan C (1 encoding byte and 3 byte checksum)

More encoding bytes offer a greater chance to parse the message and potentially avoid the work of checksum entirely, whereas lengthening the checksum and shortening the prepended encoding offers the same chance of a collision risk but with the benefit of powerful error detection - which comes at a cost of work and complexity (which i don't think is actually an issue for this particular application).

That said I have no strong feelings about which way we go as long as well specified and has access for community to add message types via github in a dedicated repo :-)
